Allow initial call to .getReference() from any thread

schmidt-sebastian opened the pull request:
This allows me to run the following in a test app without crashes:
```
  new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    final DatabaseReference reference = FirebaseDatabase.getInstance().getReference();
                    reference.child("foo/test").setValue("done");
                    } catch (Throwable t) {
                    System.out.println(t);
                    }
            }
        }).start();
```

schmidt-sebastian commented:
/test connected-check

wilhuff commented:
Could you explain a bit (or link to an explanation) about why we don't have to call `Looper.prepare()` as we used to?

schmidt-sebastian commented:
> Could you explain a bit (or link to an explanation) about why we don't have to call `Looper.prepare()` as we used to?

All the code that this PR touches was added or modified in the migration. The 16.0.1 RTDB client used com.google.android.gms.libs.punchclock.threads.TracingHandler as its thread handler, but that class is GMSCore internal. I replaced this code with Android's standard Handler interface.

In my initial work to get this to work, I ran across a wide range of issues that all boiled down to the same thing: I couldn't submit work to a Handler on a thread that hadn't be initialized with `Looper.prepare()`. I wrongfully assumed that this was a test-only problem and that any thread used in an Android app was already prepared.

So - with the fix here to use the main looper, I am only ever using threads that have already prepared. We already do this in other parts of the SDK, such as here: https://github.com/firebase/firebase-android-sdk/blob/00d46269f1ef9ce0f1dd69ab727137a84332d85c/firebase-database/src/main/java/com/google/firebase/database/android/AndroidPlatform.java#L93
Thus, I don't need the lines I added to the integration tests anymore.

That being said, I really want @bjornick to chime in on this as well.

bjornick commented:
Is there any concern about running those callbacks on the main thread?  It almost feels like we want to run that in a separate thread of our own choosing, right?

schmidt-sebastian commented:
> Is there any concern about running those callbacks on the main thread? It almost feels like we want to run that in a separate thread of our own choosing, right?

We used to expose this as an option (in the pre-GMSCore days), but since 3.x the callbacks only ever fire on the main thread.
